kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceBatchAction.java

```diff
@@ -0,0 +1,434 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collector;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.validation.Valid;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Providers;
+import org.apache.kafka.common.errors.SerializationException;
+
+@DoNotRateLimit
+@Path("/v3/clusters/{clusterId}/topics/{topicName}/records:batch")
+@ResourceName("api.v3.batch-produce.*")
+public final class ProduceBatchAction {
+
+  public static final int BATCH_ID_MINIMUM_LENGTH = 1;
+  public static final int BATCH_ID_MAXIMUM_LENGTH = 80;
+  public static final int BATCH_MAXIMUM_ENTRIES = 10;
+
+  @Context private Providers providers;
+
+  private static final Collector<
+          ProduceRequestHeader,
+          ImmutableMultimap.Builder<String, Optional<ByteString>>,
+          ImmutableMultimap<String, Optional<ByteString>>>
+      PRODUCE_REQUEST_HEADER_COLLECTOR =
+          Collector.of(
+              ImmutableMultimap::builder,
+              (builder, header) -> builder.put(header.getName(), header.getValue()),
+              (left, right) -> left.putAll(right.build()),
+              ImmutableMultimap.Builder::build);
+
+  private final Provider<SchemaManager> schemaManagerProvider;
+  private final Provider<RecordSerializer> recordSerializerProvider;
+  private final Provider<ProduceController> produceControllerProvider;
+  private final Provider<ProducerMetrics> producerMetricsProvider;
+  private final ProduceRateLimiters produceRateLimiters;
+  private final ExecutorService executorService;
+
+  @Inject
+  public ProduceBatchAction(
+      Provider<SchemaManager> schemaManagerProvider,
+      Provider<RecordSerializer> recordSerializer,
+      Provider<ProduceController> produceControllerProvider,
+      Provider<ProducerMetrics> producerMetrics,
+      ProduceRateLimiters produceRateLimiters,
+      @ProduceResponseThreadPool ExecutorService executorService) {
+    this.schemaManagerProvider = requireNonNull(schemaManagerProvider);
+    this.recordSerializerProvider = requireNonNull(recordSerializer);
+    this.produceControllerProvider = requireNonNull(produceControllerProvider);
+    this.producerMetricsProvider = requireNonNull(producerMetrics);
+    this.produceRateLimiters = requireNonNull(produceRateLimiters);
+    this.executorService = requireNonNull(executorService);
+  }
+
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  @PerformanceMetric("v3.produce.produce-to-topic")
+  @ResourceName("api.v3.produce.produce-to-topic")
```

Comment:
should this be also annotated with `batch-produce` ?

Reply:
Yes, it should. Good catch.